contracts/stakingv2/modules/StakeUnstake.sol

```diff
@@ -43,42 +42,52 @@ contract StakeUnstake is Common {
     }
 
     function _pendingSushiRewards(address _user) internal view returns (uint256) {
-        uint256 pendingSushi = IMasterChefSushi(SUSHI_MASTERCHEF).pendingSushi(OOKI_ETH_SUSHI_MASTERCHEF_PID, address(this));
-
-        uint256 totalSupply = _totalSupplyPerToken[OOKI_ETH_LP];
-        return _pendingAltRewards(SUSHI, _user, balanceOfByAsset(OOKI_ETH_LP, _user), totalSupply != 0 ? pendingSushi.mul(1e12).div(totalSupply) : 0);
+        uint256 lpBalance =  balanceOfByAsset(OOKI_ETH_LP, _user);
+        IStakingV2.AltRewardsUserInfo memory altRewardsUserInfo = userAltRewardsInfo[_user][SUSHI];
+        uint256 pendingSushi = IMasterChefSushi2(SUSHI_MASTERCHEF).pendingSushi(OOKI_ETH_SUSHI_MASTERCHEF_PID, address(this));
+        return _pendingAltRewards(SUSHI, _user, lpBalance, altRewardsUserInfo, pendingSushi);
     }
 
-
     function _pendingAltRewards(
         address token,
         address _user,
         uint256 userSupply,
-        uint256 extraRewardsPerShare
+        IStakingV2.AltRewardsUserInfo memory altRewardsUserInfo,
+        uint256 extraRewards
     ) internal view returns (uint256) {
-        uint256 _altRewardsPerShare = altRewardsPerShare[token].add(extraRewardsPerShare);
-        if (_altRewardsPerShare == 0) return 0;
+        uint256 total;
+        if(userSupply != 0){
+            uint256 _stakingStartBlock = altRewardsUserInfo.stakingStartBlock;
+            if(altRewardsUserInfo.stakingStartBlock > block.number)
+                return 0;
+
+            uint256 _altRewardsBlock  = altRewardsBlock[token]; // Last time when addAltRewards was triggered
+            uint256 _altRewardsStartBlock = altRewardsStartBlock[token];
+            if(_stakingStartBlock < _altRewardsStartBlock){
+                _stakingStartBlock = _altRewardsStartBlock;
+            }
+
+            uint256 _altRewardsPerShare;
+            uint256 _altRewardsPerSharePerBlock;
+            if(extraRewards != 0){
+                (_altRewardsPerShare, _altRewardsPerSharePerBlock, _altRewardsBlock)  = _addAltRewards(token, extraRewards);
+            }
+            else{
+                _altRewardsPerShare = altRewardsPerShare[token];
+                _altRewardsPerSharePerBlock = altRewardsPerSharePerBlock[token];
+            }
+            total = (_altRewardsBlock - _stakingStartBlock).mul(_altRewardsPerSharePerBlock).mul(userSupply).div(1e12);
```

Comment:
how do you guarantee its not possible to overflow here?

Reply:
_stakingStartBlock is <= block.number and _altRewardsBlock is set to be block.number in _addAltRewards() that is called before this